pkg/minishift/docker/image/oci_image_handler.go

```diff
@@ -0,0 +1,379 @@
+}
+
+// ImportImages imports cached images from the host into the Docker daemon of the VM.
+func (handler *OciImageHandler) ImportImages(config *ImageCacheConfig) error {
+	out := handler.getOutputWriter(config)
+	fmt.Fprintln(out, "-- Importing cached images")
+
+	policyContext, err := handler.getPolicyContext()
+	if err != nil {
+		return fmt.Errorf("Error creating security context: %s", err.Error())
+	}
+
+	availableImages, err := handler.getDockerImages()
+	if err != nil {
+		return err
+	}
+
+	for _, imageName := range config.CachedImages {
+		if _, found := availableImages[imageName]; found {
+			continue
+		}
+
+		err := handler.importImage(imageName, config, policyContext, out)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ExportImages exports the images specified as part of the ImageCacheConfig from the VM to the host.
+func (handler *OciImageHandler) ExportImages(config *ImageCacheConfig) error {
+	out := handler.getOutputWriter(config)
+
+	policyContext, err := handler.getPolicyContext()
+	if err != nil {
+		return fmt.Errorf("Error creating security context: %s", err.Error())
+	}
+
+	for _, imageName := range config.CachedImages {
+		if handler.IsImageCached(config, imageName) {
+			fmt.Fprint(out, fmt.Sprintf("%s found in cache", imageName))
+		} else {
+			err = handler.exportImage(imageName, config, policyContext, out)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsImageCached returns true if the specified image is cached, false otherwise.
+func (handler *OciImageHandler) IsImageCached(config *ImageCacheConfig, image string) bool {
+	cachedImages := handler.getCachedImages(config)
+	_, found := cachedImages[image]
+	return found
+}
+
+// AreImagesCached returns true if all images specified in the config are cached, false otherwise.
+func (handler *OciImageHandler) AreImagesCached(config *ImageCacheConfig) bool {
+	cachedImages := handler.getCachedImages(config)
+
+	for _, image := range config.CachedImages {
+		if _, found := cachedImages[image]; !found {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (handler *OciImageHandler) getCachedImages(config *ImageCacheConfig) map[string]bool {
+	cachedImages := make(map[string]bool)
+
+	index, err := handler.getIndex(config)
+	if index == nil || err != nil {
+		return cachedImages
+	}
+
+	for _, manifest := range index.Manifests {
+		cachedImages[manifest.Annotations.Name] = true
+	}
+
+	return cachedImages
+}
+
+func (handler *OciImageHandler) getDockerImages() (map[string]bool, error) {
+	dockerImages := make(map[string]bool)
+
+	session, err := handler.createSSHSession()
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	cmd := "docker images --format '{{.Repository}}:{{.Tag}}'"
+	var buffer bytes.Buffer
+	session.Stdout = &buffer
+	err = session.Run(cmd)
+	if err != nil {
+		return nil, fmt.Errorf("Error running command '%s': %v", cmd, err)
+	}
+
+	for _, image := range strings.Split(buffer.String(), "\n") {
```

Comment:
This seems to fail for me.
It seems the `buffer.String` is an empty line?

```
buffer.String() =>
dockerImages[image] => map[:true]
```

This means it will return a map with `:true` as value... and in `ImportImages` this makes `[openshift/origin:v3.6.0 openshift/origin-docker-registry:v3.6.0 openshift/origin-haproxy-router:v3.6.0]` pass as having an image cached.

Reply:
Reproducible on linux

```
$ glide cc
$ make clean
$ git pull origin
$ git checkout origin/pr/1589
$ make cross
$ minishift start
```

```
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 1% used OK
-- Importing cached images
   Importing openshift/origin:v3.6.0 . OK
-- OpenShift cluster will be configured with ...
   Version: v3.6.0
-- Checking `oc` support for startup flags ... 
   routing-suffix ... OK
   host-config-dir ... OK
   host-data-dir ... OK
   host-pv-dir ... OK
   host-volumes-dir ... OK
Starting OpenShift using openshift/origin:v3.6.0 ...
Pulling image openshift/origin:v3.6.0
Pulled 1/4 layers, 26% complete
```

Reply:
Hmm, why do you think this is an error. I agree that in the that there are no images, there will be still an execution of this loop. That's not nice and I will put a guard clause around, but I don't think it has otherwise a problem. I am not following your argument regarding _ ImportImages_.

Reply:
In this case it will do:

```
-- Importing cached images
   Importing openshift/origin:v3.6.0 . OK
```

but there aren't any images on my machine(s). Reproducible on several systems, so it is not an isolated case.